modules/unspents/src/dimensions.ts

```diff
@@ -1,287 +1,151 @@
-  getOverheadVSize(): number;
-  getInputsVSize(): number;
-  getOutputsVSize(): number;
-  getVSize(): number;
+    Object.freeze(this);
+  }
 }
 
-interface IFromInputParams {
+interface FromInputParams {
   // In cases where the input type is ambiguous, we must provide a hint about spend script type.
   assumeUnsigned?: symbol;
 }
 
-export interface IFromUnspentParams {
+export interface FromUnspentParams {
   // The p2tr output type has multiple spend options and thus different weights per spend path.
   p2trSpendType: 'keypath' | 'scriptpath-level1' | 'scriptpath-level2';
 }
 
-export interface IDimensionsStruct extends t.Struct<IDimensions> {
-  (v: IBaseDimensions): IDimensions;
-  ASSUME_P2SH: symbol;
-  ASSUME_P2SH_P2WSH: symbol;
-  ASSUME_P2WSH: symbol;
-  ASSUME_P2TR_KEYPATH: symbol;
-  ASSUME_P2TR_SCRIPTPATH_LEVEL1: symbol;
-  ASSUME_P2TR_SCRIPTPATH_LEVEL2: symbol;
-
-  SingleOutput: {
-    p2sh: IDimensions;
-    p2shP2wsh: IDimensions;
-    p2wsh: IDimensions;
-    p2tr: IDimensions;
-    p2pkh: IDimensions;
-    p2wpkh: IDimensions;
-  };
-  new (v: IBaseDimensions): IDimensions;
+/**
+ * Dimensions of a BitGo wallet transactions.
+ */
+export class Dimensions {
+  static ASSUME_P2SH = Symbol('assume-p2sh');
+  static ASSUME_P2SH_P2WSH = Symbol('assume-p2sh-p2wsh');
+  static ASSUME_P2WSH = Symbol('assume-p2wsh');
+  static ASSUME_P2TR_KEYPATH = Symbol('assume-p2tr-keypath');
+  static ASSUME_P2TR_SCRIPTPATH_LEVEL1 = Symbol('assume-p2tr-scriptpath-level1');
+  static ASSUME_P2TR_SCRIPTPATH_LEVEL2 = Symbol('assume-p2tr-scriptpath-level2');
+
+  /** Input counts for BitGo wallet multi-signature inputs */
+  public nP2shInputs = 0;
+  public nP2shP2wshInputs = 0;
+  public nP2wshInputs = 0;
+  public nP2trKeypathInputs = 0;
+  public nP2trScriptPathLevel1Inputs = 0;
+  public nP2trScriptPathLevel2Inputs = 0;
+
+  /* Input count for single-signature inputs (Replay Protection inputs) */
+  public nP2shP2pkInputs = 0;
+
+  public outputs: OutputDimensions = new OutputDimensions();
+
+  constructor(d: Partial<Dimensions> = {}) {
+    Object.entries(d).forEach(([key, value]) => this.setProperty(key, value));
+
+    Object.freeze(this);
+  }
 
-  zero(): IDimensions;
-  sum(...args: Array<Partial<IDimensions>>): IDimensions;
+  private setProperty(k: string, v: unknown): void {
+    switch (k) {
+      case 'nP2shInputs':
+      case 'nP2shP2wshInputs':
+      case 'nP2wshInputs':
+      case 'nP2trKeypathInputs':
+      case 'nP2trScriptPathLevel1Inputs':
+      case 'nP2trScriptPathLevel2Inputs':
+      case 'nP2shP2pkInputs':
+        if (typeof v !== 'number') {
+          throw new Error(`property ${k} must be number`);
+        }
+        if (!Number.isSafeInteger(v) || v < 0) {
+          throw new Error(`property ${k} must be zero or positive integer`);
+        }
+        break;
+      case 'outputs':
+        if (!(v instanceof OutputDimensions)) {
+          v = new OutputDimensions(v as OutputDimensions);
```

Comment:
Is this reachable code? Seems like it shouldn't be possible if types are properly checked. This jumped out at me because if you change `v: unknown` to `v: number | OutputDimensions` this line gives an error.

Reply:
this is to allow `new Dimensions({ outputs: { count: 1, size: 10 } })` etc